### PR TITLE
Add Hangfire instrumentation

### PR DESCRIPTION
**Pull request by fred2u**

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs:37]: `ActivityKind.Internal` would be more apt for background jobs?

File: src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs:36]: It'd be good to set displayname after the activity is confirmed to be created/sampled in, to avoid this string allocation..

File: src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs:26]: something more unique?

File: src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs:63]: `Dispose()` alone is sufficient, as it internally Stops.

File: src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs:60]: If setting status to Error, then description *must* be set as well. You may use Exception Message.

Additionally, based on opt-in option, entire  exception could be recorded as ActivityEvent as well. (Good for a separate, follow up PR, not really needed in 1st one)

File: src/OpenTelemetry.Instrumentation.Hangfire/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/README.md:7]: Consider following the structure from https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.AspNetCore

Main issue with current doc is - it does not hint that using Services.AddOpenTelemetryTracing requires additional nuget installations.

File: src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs:28]:
```suggestion
        public void OnPerforming(PerformingContext performingContext)
```

**Review by cijothomas — CHANGES_REQUESTED** [review]
Thanks for your contribution. Looks good overall, requesting changes to address some comments, and most importantly, we need to have some unit tests before merge.
(I am okay to merge if you are planning to send unit tests in a follow up PR.)

File: test/OpenTelemetry.Instrumentation.Hangfire.Tests/HangfireInstrumentationJobFilterAttributeTests.cs
Comment (cijothomas) [inline test/OpenTelemetry.Instrumentation.Hangfire.Tests/HangfireInstrumentationJobFilterAttributeTests.cs:43]: Please use InMemoryExporter in tests.
example: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3122/files

We are going to get rid of any other approaches (Like Moq). from existing tests

File: src/OpenTelemetry.Instrumentation.Hangfire/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/README.md:61]: these links are incorrect. Link to example apps: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/examples/AspNetCore/Program.cs

File: src/OpenTelemetry.Instrumentation.Hangfire/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/README.md:64]: link to example asp.net app here.

File: src/OpenTelemetry.Instrumentation.Hangfire/TracerProviderBuilderExtensions.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/TracerProviderBuilderExtensions.cs:28]:
```suggestion
        /// Adds Hangfire instrumentation to the tracer provider.
```

File: src/OpenTelemetry.Instrumentation.Hangfire/OpenTelemetry.Instrumentation.Hangfire.csproj
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/OpenTelemetry.Instrumentation.Hangfire.csproj:10]: please use the latest 1.2.0-rc4

File: src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs:57]: Use the below API to set Status
activity.SetStatus(ActivityStatusCode.Error, ex.message);

File: test/OpenTelemetry.Instrumentation.Hangfire.Tests/HangfireInstrumentationJobFilterAttributeTests.cs
Comment (cijothomas) [inline test/OpenTelemetry.Instrumentation.Hangfire.Tests/HangfireInstrumentationJobFilterAttributeTests.cs:55]: we could definitely use more tests to cover all scenarios (the one with exception etc.), but not blocking this PR for that.

File: src/OpenTelemetry.Instrumentation.Hangfire/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/README.md:31]: --prerelease

File: src/OpenTelemetry.Instrumentation.Hangfire/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/README.md:61]: OpenTelemetry.Instrumentation.AspNetCore - No need to mention this.

File: src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs:39]: is CreatedAt refering to the time the job started? This is not required, as activity already has startTime automatically captured.

File: src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs:35]:
```suggestion
            if (activity != null && activity.IsAllDataRequested)
```

**Review by cijothomas — APPROVED** [review]

**Comment by cijothomas** [comment]
Would you also update https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/.github/component_owners.yml to include your name? (It wont give you actual write/approve role to this repo, but it'll notify you for all PRs affecting this instrumentation and its tests)

File: src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs
Comment (fred2u) [reply src/OpenTelemetry.Instrumentation.Hangfire/Implementation/HangfireInstrumentationJobFilterAttribute.cs:39]: It's refering to the time the job created, not started

**Comment by cijothomas** [comment]
@fred2u Please check CI failures. Once thats green, we are good to merge!

**Comment by cijothomas** [comment]
@fred2u Thank you for your 1st contribution (of hopefully many!).

**Pull request merged**
